account_payment_mode_auto_reconcile/models/account_move.py

```diff
@@ -36,50 +34,60 @@ def invoice_validate(self):
             ).auto_reconcile_credits(partial_allowed=partial)
         return res
 
-    @api.multi
     def write(self, vals):
-        res = super(AccountInvoice, self).write(vals)
-        if "payment_mode_id" in vals:
+        res = super(AccountMove, self).write(vals)
+        if "payment_mode_id" in vals or "state" in vals:
             for invoice in self:
                 # Do not auto reconcile anything else than open customer inv
-                if invoice.state != "open" or invoice.type != "out_invoice":
+                if invoice.state != "posted" or invoice.move_type != "out_invoice":
                     continue
-                payment_mode = invoice.payment_mode_id
+                invoice_lines = invoice.line_ids.filtered(
+                    lambda line: line.account_type == "asset_receivable"
+                )
                 # Auto reconcile if payment mode sets it
+                payment_mode = invoice.payment_mode_id
                 if payment_mode and payment_mode.auto_reconcile_outstanding_credits:
                     partial = payment_mode.auto_reconcile_allow_partial
                     invoice.with_context(
                         _payment_mode_auto_reconcile=True
                     ).auto_reconcile_credits(partial_allowed=partial)
                 # If the payment mode is not using auto reconcile we remove
                 #  the existing reconciliations
-                elif invoice.payment_move_line_ids:
+                elif any(
+                    [
+                        invoice_lines.mapped("matched_credit_ids"),
+                        invoice_lines.mapped("matched_debit_ids"),
+                    ]
+                ):
                     invoice.auto_unreconcile_credits()
         return res
 
-    @api.multi
     def auto_reconcile_credits(self, partial_allowed=True):
         for invoice in self:
-            if not invoice.has_outstanding:
+            invoice._compute_payments_widget_to_reconcile_info()
```

Comment:
It seems weird to me that you have to explicitely call this before accessing the value :confused: